### ISSUE TITLE
Training progress

### ISSUE THREAD

**Issue by fhrzn** [fhrzn/detox-t5vae#2]
Will update training histories (logs, plots, configs, problems, etc.) here

**Comment by fhrzn** [comment]
For training progress live updates [wandb](https://wandb.ai/fhrzn/cycle-trans-2?workspace=user-fhrzn)

**Comment by fhrzn** [comment]
Results after training for 50 epochs (~1d 20h roughly)

<img width="1089" alt="image" src="https://user-images.githubusercontent.com/28261222/218426395-6dee7541-215d-460e-8ef1-9528b9ffead5.png">
<img width="1089" alt="image" src="https://user-images.githubusercontent.com/28261222/218426991-c2314bf7-f471-4afc-ab9f-8f9f7e87533b.png">

After some epochs we can see generator loss start to decrease and discriminator start to increase (by very little magnitude). But the progress is very slow, probably due to the learning rate.

Config:
```python3
# dataset config"
'seq_length': 128,
'tokenizer_ckpt': 'bert-base-uncased',
'batch_size': 32,
# discriminator config
'disc_d_model': 384,
'disc_num_head': 6,
'disc_num_layers': 2,
'disc_dropout': 0.2,
# generator config
'gen_d_model': 384,
'gen_num_head': 6,
'gen_num_layers': 2,
'gen_dropout': 0.5,
# optimizer config
'learning_rate': 1e-5,
'beta1': 0.5,
'beta2': 0.999,
# training config
'num_epoch': 50,
'lambda_cyc': 1,
'lambda_ident': 1,
'lambda_adv': 3
```

**Comment by fhrzn** [comment]
### Changelogs:

- Added gumbel temperature (linearly increased, following RelGAN implementation)
- Changed weight of `cycle`, `identity`, and `adv` loss

![image](https://user-images.githubusercontent.com/28261222/220822249-b88f1922-ac24-45e8-8f07-7bb47d1bf7d8.png)
![image](https://user-images.githubusercontent.com/28261222/220822342-2f7e0391-38fa-4327-880c-3358b9c19996.png)

Generator faced much more hard times to beat discriminator than previous settings. Also detoxification result still inappropriate.

Config:
```python3
config = {
    # dataset config"
    'seq_length': 128,
    'tokenizer_ckpt': 'bert-base-uncased',
    'batch_size': 8,
    # discriminator config
    'disc_d_model': 384,
    'disc_num_head': 6,
    'disc_num_layers': 2,
    'disc_dropout': 0.2,
    # generator config
    'gen_d_model': 384,
    'gen_num_head': 6,
    'gen_num_layers': 2,
    'gen_dropout': 0.5,
    # optimizer config
    'learning_rate': 1e-5,
    'beta1': 0.5,
    'beta2': 0.999,
    # training config
    'num_epoch': 50,
    'lambda_cyc': 1,
    'lambda_ident': 1,
    'lambda_adv': 3,
    'gumbel_temp': 1000
}

**Comment by fhrzn** [comment]
### Changelogs:

- Revert gumbel temperature (for now)
- Readjust losses weight and learning rate (following [eliorav](https://github.com/eliorav/writing-style-transfer))

![image](https://user-images.githubusercontent.com/28261222/220822806-d0ee8ef9-857f-48b5-8ac1-25705ef63381.png)
![image](https://user-images.githubusercontent.com/28261222/220822869-3dd2877b-50a0-4abe-8355-cdc77a503972.png)

Still, generator cannot beat discriminator.

Config:
```python3
config = {
    # dataset config"
    'seq_length': 128,
    'tokenizer_ckpt': 'bert-base-uncased',
    'batch_size': 8,
    # discriminator config
    'disc_d_model': 384,
    'disc_num_head': 6,
    'disc_num_layers': 2,
    'disc_dropout': 0.2,
    # generator config
    'gen_d_model': 384,
    'gen_num_head': 6,
    'gen_num_layers': 2,
    'gen_dropout': 0.5,
    # optimizer config
    'learning_rate': 2e-4,
    'beta1': 0.5,
    'beta2': 0.999,
    # training config
    'num_epoch': 50,
    'lambda_cyc': 8,
    'lambda_ident': 5,
    'lambda_adv': 1,
}

**Comment by fhrzn** [comment]
### Changelogs:

- Trying to change discriminator role as toxic classifier
- Discriminator objective is to guide generator by giving label prediction of generated sentences
- Pretrained toxic classifier as discriminator
- Discriminator weights' are freezed during generator training

![image](https://user-images.githubusercontent.com/28261222/220823708-cbcb3bcf-c0f3-4315-9a05-3a470aeda65f.png)
![image](https://user-images.githubusercontent.com/28261222/220823733-bed35fce-17e4-40b8-a750-ed1ad1d41278.png)

The generator seems to ignore discriminator label (or probably I got wrong implementation here). The common problem still occur: the detoxified sentences seems like copying the input sentences (exactly the same).

Config:
```python3
# wandb config
config = {
    # dataset config"
    'seq_length': 128,
    'tokenizer_ckpt': 'bert-base-uncased',
    'batch_size': 32,
    # discriminator config
    'disc_d_model': 512,
    'disc_num_head': 8,
    'disc_num_layers': 6,
    'disc_dropout': 0.2,
    # generator config
    'gen_d_model': 384,
    'gen_num_head': 6,
    'gen_num_layers': 2,
    'gen_dropout': 0.5,
    # optimizer config
    'learning_rate': 2e-4,
    'beta1': 0.5,
    'beta2': 0.999,
    # training config
    'num_epoch': 50,
    'lambda_cyc': 5,
    'lambda_ident': 2,
    'lambda_adv': 8
}

**Comment by fhrzn** [comment]
Just figured out that the gradient might be not flowing from the discriminator to the generator. Possibly it causes the main issue where the generation output always identically to the input (bcs the adversarial loss didn't contribute anything).

For now, I used Linear layer to embed generator logits to discriminator since turns out the gradient cannot flow through Embedding layer too ([ref](https://discuss.pytorch.org/t/differentiable-argmax/33020/9?u=fhrzn), [ref](https://discuss.pytorch.org/t/predict-a-categorical-variable-and-then-embed-it-one-hot/85555/3))

Another useful [reference](https://wiki.lzhbrian.me/notes/differientiable-sampling-and-argmax) and [reference](https://discuss.pytorch.org/t/optimizing-based-on-another-models-output/6935/19).

**Comment by fhrzn** [comment]
The common problems that we always mentioned about
<img width="631" alt="image" src="https://user-images.githubusercontent.com/28261222/222147902-21c44e91-11c6-44c9-a909-26e57fd8406f.png">

**Comment by fhrzn** [comment]
We then decided to perform experiments to identify what parts that make the common problems happened. Initially, we suggest either cycle loss or identity loss that triggers the problem. So, we will try to turn off one of them in each experiment.
_Note that to investigate the problem, we will observe losses in each step._

**Experiment result without identity loss**
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/28261222/222149432-0bd38f73-a9bb-4820-abaf-2f4b14f10d12.png">
<img width="610" alt="image" src="https://user-images.githubusercontent.com/28261222/222151382-3d26f3ba-6287-4d93-905a-74ac7daa966f.png">


**Experiment result without cycle loss**
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/28261222/222149525-7bc1cdc2-1062-42fc-bb8f-6eaaaa36cfa0.png">
<img width="610" alt="image" src="https://user-images.githubusercontent.com/28261222/222150967-80c60f40-b8f1-44de-9d77-503680d634ed.png">


It can be seen that identity loss plays an important role here. Without identity loss, the cycle of loss seems to be suffering to translate the sentence. Otherwise, without cycle loss, identity loss doesn't have any problem at all. We also observed that the generation result of the experiment without identity loss is just random sentences. On the other hand, the generation result without cycle loss is very close to the original input sentences. Therefore, we conclude this identity loss has a lot of power to trigger common problems to occur.

**Comment by fhrzn** [comment]
In the @efd4152, we attempt to pre-train the generator with the identity loss objective. We trained it to have not so low loss value, so the model has an initial assumption about the data but is not so perfect. We then load it as a pre-trained generator. Finally, we turned off the identity loss objective in the current training progress.

**Comment by fhrzn** [comment]
seems all this time I wrongly implemented the Generator sample function. will do further check tomorrow
